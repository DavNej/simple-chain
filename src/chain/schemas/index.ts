import { z } from 'zod'

export const AddressSchema = z.string().regex(/^0x[a-fA-F0-9]{40}$/)
