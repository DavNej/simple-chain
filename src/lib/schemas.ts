import { z } from 'zod'

export const addressRegex = /^0x[a-fA-F0-9]{40}$/
export const keccak256Regex = /^0x[a-fA-F0-9]{64}$/

export const AddressSchema = z
  .string()
  .regex(addressRegex, { message: 'Wrong format for address' })

export const MessageSchema = z.coerce
  .string()
  .max(260, {
    message: "Message can't exceed 260 characters",
  })
  .optional()

export const keccak256Schema = z
  .string()
  .regex(keccak256Regex, { message: 'Wrong format for hash' })
