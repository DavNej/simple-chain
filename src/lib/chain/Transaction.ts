import { keccak256, toUtf8Bytes } from 'ethers'
import { stringToHex } from '@/lib/utils'
import type { TransactionArgsType } from './types'

export default class Transaction {
  from: string
  to: string
  value: number
  data: string | null
  message: string | null
  createdAt: number
  hash: string

  constructor({ from, to, value, data, message }: TransactionArgsType) {
    this.from = from
    this.to = to
    this.value = value || 0
    this.data = data ? stringToHex(data) : null
    this.message = message || null
    this.createdAt = Date.now()
    this.hash = this.calculateHash()
  }

  calculateHash() {
    return keccak256(
      toUtf8Bytes(
        [
          this.data,
          this.from,
          this.to,
          this.value,
          this.createdAt,
          this.message,
        ].join(''),
      ),
    )
  }

  verify() {
    return this.hash === this.calculateHash()
  }
}
