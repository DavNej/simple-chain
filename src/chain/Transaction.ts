import type { AddressType } from './types'
import type { MessageType } from './types/transaction'
import { TransactionSchema } from './schemas/transaction'

export class Transaction {
  from: AddressType
  to: AddressType
  amount: number
  data: unknown
  message: MessageType

  constructor({
    from,
    to,
    amount,
    data,
    message,
  }: {
    from: AddressType
    to: AddressType
    amount: number
    data: unknown
    message: MessageType
  }) {
    TransactionSchema.parse({ from, to, amount, data, message })

    this.from = from
    this.to = to
    this.amount = amount
    this.data = data
    this.message = message
  }
}
