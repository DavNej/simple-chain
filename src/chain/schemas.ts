import { z } from 'zod'
import Transaction from './Transaction'

export const addressRegex = /^0x[a-fA-F0-9]{40}$/
export const SHA256Regex = /^[a-fA-F0-9]{64}$/

export const AddressSchema = z.string().regex(addressRegex)
export const MessageSchema = z.coerce.string().max(260).default('')

export const BlockArgsSchema = z.object({
  index: z.number().int().nonnegative(),
  difficulty: z.number().int().nonnegative().lte(64).default(0),
  prevHash: z.string().regex(SHA256Regex),
  message: MessageSchema,
  transactions: z.instanceof(Transaction).array(),
})

export const TransactionArgsSchema = z.object({
  from: AddressSchema,
  to: AddressSchema,
  value: z.number().int().lte(1e6).nonnegative(),
  data: z.unknown(),
  message: MessageSchema,
})
