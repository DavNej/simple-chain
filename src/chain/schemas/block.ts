import { z } from 'zod'
import { MessageSchema, SHA256Schema } from '../schemas'
import { TransactionListSchema } from './transaction'

export const BlockSchema = z.object({
  index: z.number().int().nonnegative(),
  timeStamp: z.date(),
  difficulty: z.number().int().nonnegative().lte(64).default(0),
  prevHash: SHA256Schema,
  hash: SHA256Schema,
  nonce: z.number().int().nonnegative(),
  message: MessageSchema,
  transactions: TransactionListSchema,
})

export const BlockListSchema = BlockSchema.array()
