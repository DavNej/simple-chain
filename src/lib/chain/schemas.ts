import * as z from 'zod'
import Transaction from './transaction'
import { AddressSchema, MessageSchema, SHA256Schema } from '@/lib/schemas'

export const transactionArgsSchema = z.object({
  from: AddressSchema,
  to: AddressSchema,
  value: z.coerce.number().int().lte(1e6).nonnegative().default(0),
  data: z.string().optional(),
  message: MessageSchema,
})

export const blockArgsSchema = z.object({
  index: z.number().int().nonnegative(),
  difficulty: z.number().int().nonnegative().lte(32).default(0),
  prevHash: SHA256Schema,
  message: MessageSchema,
  transactions: z.instanceof(Transaction).array().nonempty(),
})
