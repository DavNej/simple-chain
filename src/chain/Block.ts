import { SHA256 } from 'crypto-js'

import type { SHA256Type, MessageType } from './types'
import type { TransactionListType } from './types/transaction'
import type { BlockType } from './types/block'

import { BlockSchema } from './schemas/block'

export default class Block implements BlockType {
  index: number
  timeStamp: Date
  difficulty: number
  prevHash: SHA256Type
  hash: SHA256Type
  nonce: number
  message: MessageType
  transactions: TransactionListType

  constructor({
    index,
    difficulty,
    prevHash,
    message,
    transactions,
  }: {
    index: number
    difficulty: number
    prevHash: SHA256Type
    transactions: TransactionListType
    message: MessageType
  }) {
    this.index = index
    this.timeStamp = new Date()
    this.difficulty = difficulty
    this.prevHash = prevHash
    this.nonce = 0
    this.message = message
    this.transactions = transactions
    this.hash = this.computeHash()

    BlockSchema.parse(this)
  }

  computeHash() {
    const hash = SHA256(
      [
        this.index,
        this.timeStamp,
        this.prevHash,
        JSON.stringify(this.transactions),
        this.nonce,
        this.message,
      ].join('')
    ).toString()

    this.hash = hash
    return this.hash
  }

  async mine() {
    console.log('⛏️ Mining block ...')

    return new Promise(resolve => {
      while (!this.hash.startsWith('0'.repeat(this.difficulty))) {
        this.nonce++
        this.computeHash()
      }

      console.log('⛏️ Block mined ! ✅', this.hash)

      resolve(this.hash)
    })
  }
}
