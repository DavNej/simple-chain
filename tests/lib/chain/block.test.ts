import { waitFor } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import Block from '@/lib/chain/block'
import Transaction from '@/lib/chain/transaction'
import { keccak256Regex } from '@/lib/schemas'
import { buildBlock, mock } from '@/tests/test-utils/helpers'

let block: Block
let minedBlock: Block
let corruptedBlock: Block

beforeAll(async () => {
  vi.useFakeTimers().setSystemTime(new Date(mock.SYSTEM_DATE))
  block = buildBlock()

  minedBlock = buildBlock()
  await minedBlock.mine()

  corruptedBlock = buildBlock()
  await corruptedBlock.mine()
  corruptedBlock.transactions[0].value += 52
  vi.useRealTimers()
})

describe('Block Class', () => {
  it('creates a valid block', () => {
    expect(block).toBeInstanceOf(Block)
    expect(block.index).toBe(mock.BLOCK_ARGS.index)
    expect(block.difficulty).toBe(mock.BLOCK_ARGS.difficulty)
    expect(block.prevHash).toBe(mock.BLOCK_ARGS.prevHash)
    expect(block.message).toBe(mock.BLOCK_ARGS.message)
    expect(block.createdAt).toBe(mock.SYSTEM_TIMESTAMP)
    expect(block.hash).toBeNull()
    expect(block.nonce).toBeNull()

    expect(block.transactions).toHaveLength(2)
    expect(block.transactions[0]).toBeInstanceOf(Transaction)
    expect(block.transactions[0].createdAt).toBe(mock.SYSTEM_TIMESTAMP)
    expect(block.transactions[1]).toBeInstanceOf(Transaction)
  })

  it('should correctly calculate merkel tree', () => {
    const tree = block.calculateMerkelTree()
    expect(tree).toMatchSnapshot()
  })

  it('should correctly calculate merkel root', () => {
    const merkelRoot = block.calculateMerkelRoot()
    expect(merkelRoot).toMatchInlineSnapshot(
      `"859f814d92737c845b5a4d9c548a249def950abb1e0b7751960d90101a2157df"`,
    )
  })

  it('should correctly calculate hash', () => {
    expect(() => block.calculateHash(null)).toThrowErrorMatchingInlineSnapshot(
      `[Error: nonce must be given a numerical value]`,
    )
    expect(block.calculateHash(1)).toMatchInlineSnapshot(
      `"0x28408edf8d3fa8459741b397e5d3a2e3b7a6a2aa87feb0222aab59e570d58e17"`,
    )
    expect(block.hash).toBeNull() // only mining a block assigns a hash to it
    expect(block.nonce).toBeNull() // only mining a block assigns a nonce to it
  })

  it('mines a block', () => {
    waitFor(async () => {
      const hash = await block.mine()
      expect(block.hash).toMatch(keccak256Regex)
      expect(block.hash).toBe(hash)
      expect(block.hash!.startsWith('0'.repeat(block.difficulty))).toBe(true)
      expect(block.nonce).toBeInstanceOf(Number)
    })
  })

  it('should verify transactions', () => {
    const isValid = minedBlock.verifyTransactions()
    expect(isValid).toBe(true)
  })

  it('should validate itself', () => {
    const isValid = minedBlock.verify()
    expect(isValid).toBe(true)
  })

  it('should fail to verify transactions of a corrupted block', () => {
    const isValid = corruptedBlock.verifyTransactions()
    expect(isValid).toBe(false)
  })

  it('should fail to mine a mined block', () => {
    expect(async () => await minedBlock.mine()).rejects.toMatchInlineSnapshot(
      `"Block has already been mined"`,
    )
  })

  it('should fail to validate a corrupted block', () => {
    const isValid = corruptedBlock.verify()
    expect(isValid).toBe(false)
  })

  it('should fail to match merkle tree of a corrupted block', () => {
    const tree = corruptedBlock.calculateMerkelTree()
    expect(tree).not.toMatchSnapshot()
  })

  it('should fail to match merkle root of a corrupted block', () => {
    const merkelRoot = block.calculateMerkelRoot()
    const merkelRootCorrupted = corruptedBlock.calculateMerkelRoot()

    expect(merkelRoot).toMatchInlineSnapshot(
      `"859f814d92737c845b5a4d9c548a249def950abb1e0b7751960d90101a2157df"`,
    )
    expect(merkelRootCorrupted).not.toBe(merkelRoot)
  })
})
