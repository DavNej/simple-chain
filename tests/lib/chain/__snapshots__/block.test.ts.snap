// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Block Class > should correctly calculate merkel tree 1`] = `
MerkleTree {
  "complete": false,
  "concatenator": [Function],
  "duplicateOdd": false,
  "fillDefaultHash": null,
  "hashFn": [Function],
  "hashLeaves": false,
  "isBitcoinTree": false,
  "layers": [
    [
      {
        "data": [
          109,
          229,
          248,
          234,
          137,
          1,
          211,
          169,
          171,
          205,
          87,
          157,
          44,
          195,
          53,
          64,
          158,
          104,
          156,
          234,
          29,
          131,
          38,
          250,
          248,
          143,
          193,
          228,
          16,
          219,
          78,
          148,
        ],
        "type": "Buffer",
      },
      {
        "data": [
          70,
          235,
          73,
          13,
          108,
          64,
          210,
          166,
          204,
          68,
          10,
          49,
          39,
          39,
          79,
          175,
          51,
          95,
          248,
          102,
          161,
          48,
          89,
          139,
          46,
          13,
          82,
          91,
          66,
          221,
          86,
          250,
        ],
        "type": "Buffer",
      },
    ],
    [
      {
        "data": [
          10,
          171,
          206,
          64,
          159,
          66,
          125,
          170,
          188,
          76,
          184,
          7,
          62,
          218,
          3,
          174,
          242,
          237,
          124,
          2,
          122,
          218,
          46,
          241,
          34,
          181,
          206,
          195,
          126,
          155,
          38,
          40,
        ],
        "type": "Buffer",
      },
    ],
  ],
  "leaves": [
    {
      "data": [
        109,
        229,
        248,
        234,
        137,
        1,
        211,
        169,
        171,
        205,
        87,
        157,
        44,
        195,
        53,
        64,
        158,
        104,
        156,
        234,
        29,
        131,
        38,
        250,
        248,
        143,
        193,
        228,
        16,
        219,
        78,
        148,
      ],
      "type": "Buffer",
    },
    {
      "data": [
        70,
        235,
        73,
        13,
        108,
        64,
        210,
        166,
        204,
        68,
        10,
        49,
        39,
        39,
        79,
        175,
        51,
        95,
        248,
        102,
        161,
        48,
        89,
        139,
        46,
        13,
        82,
        91,
        66,
        221,
        86,
        250,
      ],
      "type": "Buffer",
    },
  ],
  "sort": false,
  "sortLeaves": false,
  "sortPairs": false,
}
`;

exports[`Block Class > should fail to match merkle tree of a corrupted block 1`] = `
MerkleTree {
  "complete": false,
  "concatenator": [Function],
  "duplicateOdd": false,
  "fillDefaultHash": null,
  "hashFn": [Function],
  "hashLeaves": false,
  "isBitcoinTree": false,
  "layers": [
    [
      {
        "data": [
          109,
          229,
          248,
          234,
          137,
          1,
          211,
          169,
          171,
          205,
          87,
          157,
          44,
          195,
          53,
          64,
          158,
          104,
          156,
          234,
          29,
          131,
          38,
          250,
          248,
          143,
          193,
          228,
          16,
          219,
          78,
          148,
        ],
        "type": "Buffer",
      },
      {
        "data": [
          70,
          235,
          73,
          13,
          108,
          64,
          210,
          166,
          204,
          68,
          10,
          49,
          39,
          39,
          79,
          175,
          51,
          95,
          248,
          102,
          161,
          48,
          89,
          139,
          46,
          13,
          82,
          91,
          66,
          221,
          86,
          250,
        ],
        "type": "Buffer",
      },
    ],
    [
      {
        "data": [
          10,
          171,
          206,
          64,
          159,
          66,
          125,
          170,
          188,
          76,
          184,
          7,
          62,
          218,
          3,
          174,
          242,
          237,
          124,
          2,
          122,
          218,
          46,
          241,
          34,
          181,
          206,
          195,
          126,
          155,
          38,
          40,
        ],
        "type": "Buffer",
      },
    ],
  ],
  "leaves": [
    {
      "data": [
        109,
        229,
        248,
        234,
        137,
        1,
        211,
        169,
        171,
        205,
        87,
        157,
        44,
        195,
        53,
        64,
        158,
        104,
        156,
        234,
        29,
        131,
        38,
        250,
        248,
        143,
        193,
        228,
        16,
        219,
        78,
        148,
      ],
      "type": "Buffer",
    },
    {
      "data": [
        70,
        235,
        73,
        13,
        108,
        64,
        210,
        166,
        204,
        68,
        10,
        49,
        39,
        39,
        79,
        175,
        51,
        95,
        248,
        102,
        161,
        48,
        89,
        139,
        46,
        13,
        82,
        91,
        66,
        221,
        86,
        250,
      ],
      "type": "Buffer",
    },
  ],
  "sort": false,
  "sortLeaves": false,
  "sortPairs": false,
}
`;
